sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -589,6 +594,9 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
       IntervalUtils.castStringToDTInterval(s, it.startField, it.endField))
     case _: DayTimeIntervalType => buildCast[Long](_, s =>
       IntervalUtils.durationToMicros(IntervalUtils.microsToDuration(s), it.endField))
+    case x: IntegralType if it.startField == it.endField =>
+      b => IntervalUtils.longToDayTimeInterval(
+        x.exactNumeric.asInstanceOf[Numeric[Any]].toLong(b), it.endField)
```

Comment:
Why don't you handle overflow here like for year-month intervals?

Reply:
No overflow is possible that onvert byte/short/int/long to long. It's not like year-month intervals, because there is a case that long to int.